packages/patrol_cli/lib/src/coverage/device_to_host_port_transformer.dart

```diff
@@ -0,0 +1,80 @@
+import 'dart:async';
+
+import 'package:adb/adb.dart';
+import 'package:patrol_cli/src/base/logger.dart';
+import 'package:patrol_cli/src/coverage/vm_connection_details.dart';
+import 'package:patrol_cli/src/devices.dart';
+import 'package:process/process.dart';
+
+class DeviceToHostPortTransformer
+    implements StreamTransformer<VMConnectionDetails, VMConnectionDetails> {
+  DeviceToHostPortTransformer({
+    required ProcessManager processManager,
+    required TargetPlatform devicePlatform,
+    required Adb adb,
+    required Logger logger,
+  })  : _processManager = processManager,
+        _devicePlatform = devicePlatform,
+        _adb = adb,
+        _logger = logger;
+
+  final ProcessManager _processManager;
+  final TargetPlatform _devicePlatform;
+  final Adb _adb;
+  final Logger _logger;
+
+  /// The number was chosen randomly
+  static const _hostPort = 61011;
+
+  @override
+  Stream<VMConnectionDetails> bind(Stream<VMConnectionDetails> stream) async* {
+    await for (final value in stream) {
+      final hostPort = await _devicePortToHostPort(value.port);
+      if (hostPort != null) {
+        yield VMConnectionDetails(
+          port: hostPort,
+          auth: value.auth,
+        );
+      }
+    }
+  }
+
+  Future<int?> _devicePortToHostPort(int devicePort) async {
+    final int? hostPort;
+
+    switch (_devicePlatform) {
+      case TargetPlatform.android:
+        await _adb.forwardPorts(fromHost: _hostPort, toDevice: devicePort);
+
+        // It is necessary to grab the port from adb forward --list because
+        // if debugger was attached, the port might be different from the one
+        // we set
+        final forwardList = await _processManager.run(
+          // TODO: Add to wrapper
+          ['adb', 'forward', '--list'],
+          runInShell: true,
+        );
+        final output = forwardList.stdout as String;
+        hostPort = int.tryParse(
+          RegExp('tcp:([0-9]+) tcp:$devicePort').firstMatch(output)?.group(1) ??
+              '',
+        );
```

Comment:
Be careful with calling `adb` directly – sometimes the adb server might not be up, and then it will print more than just output of `adb forward --list`. For example:

```
$ adb kill-server # explicitly kill adb server
$ adb forward --list
* daemon not running; starting now at tcp:5037
* daemon started successfully
```

I think the above output will make `int.tryParse` return null.

It might not be the case here (`flutter run` (or equivalent) had likely been execute before, so adb connection is there), but let's be on the safe side w.r.t input validation.

(btw, `adb forward` sounds like it should be added to `package:adb`, but i don't wanna complain too much)

Reply:
afair, if the `adb` server is not running, and we start it by calling `adb forward --list`, the list will be empty anyway. What they say [here](https://developer.android.com/studio/run/emulator-networking#adbredir) implies that stopping adb server also stops the redirections. Therefore, it seems reasonable to me that we return null in this case

Reply:
if it did return a proper output after a few lines of startup logs, we would parse the ports properly, wouldn't we? the regex takes the whole output and searches for a match, so I think that would not be an issue

Reply:
if you tested that situation and it works, then that's cool. i just wanted to make sure it's handled.

btw, in the future I'd really advise to implement this "adb forward" functionality in your `adb` package, now i remembered it has some [additional checks](https://github.com/leancodepl/patrol/blob/9db9449afe2410b221ef61e890d6b9ba83889ada/packages/adb/lib/src/internals.dart#L33-L98) that are there for a reason.

Reply:
added in #2332